### PR TITLE
Bug/591 slicing memory issues

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

`getitem`/`setitem` now use tuples for most key types. 
Advanced indexing is also implemented via this approach. keys for advanced indexing can be `lists`, `torch.Tensors`, or `ht.DNDarrays`. This means that the results of `where` and `nonzero` can be fed directly into another DNDarray as a key. However, it also means that `list`s are treated differently than `tuple`s when given as a key

Issue/s resolved: #591 #505 

## Changes proposed:
- most keys in `getitem`/`setitem` create a tuple for a given key
- removal of other types of key options in `getitem`/`setitem`
- `reduce_op` neutral value array created using the gshape instead of the lshape
- addition of advanced indexing

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?

yes, any function which uses `where` or `nonzero` can now use the DNDarray instead of a list.

### CODE REVIEW

File: heat/core/dndarray.py [hunk 1]
```diff
@@ -1399,222 +1401,175 @@ def __getitem__(self, key):
         (1/2) >>> tensor([0.])
         (2/2) >>> tensor([0., 0.])
         """
-
+        # print('start getitem', key)
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py:1404]: Remove superfluous lines

File: heat/core/dndarray.py [hunk 2]
```diff
@@ -1399,222 +1401,175 @@ def __getitem__(self, key):
...
+        # print()
         l_dtype = self.dtype.torch_type()
-        if isinstance(key, DNDarray) and key.gshape[-1] != len(self.gshape):
-            key = tuple(x.item() for x in key)
+        kgshape_flag = False
+        if isinstance(key, DNDarray) and key.ndim == self.ndim:
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py:1408]: Would you mind to add some high-level comments here as to what each of these branches are doing?

File: heat/core/dndarray.py [hunk 3]
```diff
@@ -1399,222 +1401,175 @@ def __getitem__(self, key):
...
+            if isinstance(k, slice):
+                new_slice = stride_tricks.sanitize_slice(k, self.gshape[c])
+                gout_full[c] = math.ceil((new_slice.stop - new_slice.start) / new_slice.step)
+            elif isinstance(k, list):
+                gout_full[c] = len(k)
+            elif isinstance(k, (DNDarray, torch.Tensor)):
+                gout_full[c] = k.shape[0] if not kgshape_flag else kgshape[c]
+            if isinstance(k, DNDarray):
+                key[c] = k._DNDarray__array
+        if all([g == 1 for g in gout_full]):
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py:1442]: remove the square brackets, then the expression will become a generator and not copy to list afterwards

File: heat/core/dndarray.py [hunk 4]
```diff
@@ -1399,222 +1401,175 @@ def __getitem__(self, key):
...
+            ends = torch.tensor(ends, device=self.device.torch_device)
+            chunk_ends = ends.cumsum(dim=0)
+            chunk_starts = torch.tensor([0] + chunk_ends.tolist(), device=self.device.torch_device)
+            chunk_start = chunk_starts[rank]
+            chunk_end = chunk_ends[rank]
+            arr = torch.Tensor()
+            # all keys should be tuples here
+            gout = [0] * len(self.gshape)
+            # handle the dimensional reduction for integers
+            ints = sum([isinstance(it, int) for it in key])
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py:1485]: probably this should be an all() instead of sum(), also: remove the square brackets and transform into a generator

File: heat/core/dndarray.py [hunk 5]
```diff
@@ -1399,222 +1401,175 @@ def __getitem__(self, key):
...
+                )
 
-                gout = list(self.__array[key].shape)
-                arr = self.__array[key]
+                loc_inds = torch.where((inds >= chunk_start) & (inds < chunk_end))
+                if len(loc_inds[0]) != 0:
+                    inds = inds[loc_inds] - chunk_start
+                    lkey[self.split] = inds
+                    arr = self.__array[tuple(lkey)]
+            elif isinstance(key[self.split], slice):
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py:1517]: Can we have some high-level comments here as well?

File: heat/core/stride_tricks.py [hunk 6]
```diff
@@ -158,3 +158,31 @@ def sanitize_shape(shape):
...
     return shape
+
+
+def sanitize_slice(sl, max_dim) -> slice:
+    """
+    Remove None-types from a slice
+
+    Parameters
+    ----------
+    sl : slice
```
Comment (Markus-Goetz) [inline heat/core/stride_tricks.py:169]: Parameter explanation is missing

File: heat/core/stride_tricks.py [hunk 7]
```diff
@@ -158,3 +158,31 @@ def sanitize_shape(shape):
...
+        maximum index for the given slice
+
+    Raises
+    ------
+    TypeError
+        if sl is not a slice
+    """
+    if not isinstance(sl, slice):
+        raise TypeError("This function is only for slices!")
+    new_sl = [None] * 3
```
Comment (Markus-Goetz) [inline heat/core/stride_tricks.py:180]: line break

File: heat/core/stride_tricks.py [hunk 8]
```diff
@@ -158,3 +158,31 @@ def sanitize_shape(shape):
...
+    new_sl[0] = 0 if sl.start is None else sl.start
+    if new_sl[0] < 0:
+        new_sl[0] += max_dim
+    new_sl[1] = max_dim if sl.stop is None else sl.stop
+    if new_sl[1] < 0:
+        new_sl[1] += max_dim
+    new_sl[2] = 1 if sl.step is None else sl.step
+    return slice(new_sl[0], new_sl[1], new_sl[2])
```
Comment (Markus-Goetz) [inline heat/core/stride_tricks.py:188]: line break

File: heat/core/tests/test_dndarray.py [hunk 9]
```diff
@@ -778,6 +778,7 @@ def test_setitem_getitem(self):
 
         a = ht.zeros((13, 5), split=0, device=ht_device)
         a[-1] = 1
+        # print('here', a)
```
Comment (Markus-Goetz) [inline heat/core/tests/test_dndarray.py:781]: Remove dead code

File: heat/core/tests/test_dndarray.py [hunk 10]
```diff
@@ -961,8 +962,12 @@ def test_setitem_getitem(self):
         # setting with heat tensor
         a = ht.zeros((4, 5), split=1, device=ht_device)
         a[1, 0:4] = ht.arange(4, device=ht_device)
+        # print(a)
```
Comment (Markus-Goetz) [inline heat/core/tests/test_dndarray.py:965]: Remove dead code

File: heat/core/dndarray.py (hunk 4 above)
Comment (coquelin77) [reply heat/core/dndarray.py:1485]: in this case, the number of integers in the key is used to determine the new split axis so it as to be the sum. but the square brackets were removed
